compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUTileTensor.cpp

```diff
@@ -0,0 +1,183 @@
+                                                    filter);
+}
+
+static LogicalResult tileReduction(func::FuncOp funcOp) {
+  {
+    // Tile again at the workgroup level since redution dimension were
+    // ignored. Dimensions already tiled will be ignore since we tile to the
+    // same size.
+    RewritePatternSet wgTilingPatterns(funcOp.getContext());
+    populateTilingReductionPatterns(wgTilingPatterns);
+    if (failed(applyPatternsAndFoldGreedily(funcOp,
+                                            std::move(wgTilingPatterns)))) {
+      return failure();
+    }
+  }
+
+  {
+    RewritePatternSet wgTilingCanonicalizationPatterns =
+        linalg::getLinalgTilingCanonicalizationPatterns(funcOp.getContext());
+    populateAffineMinSCFCanonicalizationPattern(
+        wgTilingCanonicalizationPatterns);
+    if (failed(applyPatternsAndFoldGreedily(
+            funcOp, std::move(wgTilingCanonicalizationPatterns)))) {
+      return failure();
+    }
+    return success();
+  }
+}
+
+/// Tile parallel dimensions according to the attribute tile sizes attached to
+/// each op.
+static LogicalResult tileParallelDims(func::FuncOp funcOp,
+                                      SmallVectorImpl<int64_t> &workgroupSize,
+                                      bool distributeToWarp) {
+  std::array<int64_t, 3> elementPerWorkgroup = {
+      distributeToWarp ? workgroupSize[0] / kWarpSize : workgroupSize[0],
+      workgroupSize[1], workgroupSize[2]};
+  SmallVector<Operation *> computeOps;
+  SmallVector<LoopTilingAndDistributionInfo> tiledLoops;
+  if (failed(getComputeOps(funcOp, computeOps, tiledLoops))) {
+    return funcOp.emitOpError("failed to get compute ops");
+  }
+
+  for (Operation *op : computeOps) {
+    auto tilingOp = dyn_cast<TilingInterface>(op);
+    if (!tilingOp) continue;
+    size_t numLoops = 0;
+    for (auto type : tilingOp.getLoopIteratorTypes()) {
+      if (type == getParallelIteratorTypeName()) numLoops++;
+    }
+    IRRewriter rewriter(op->getContext());
+    rewriter.setInsertionPoint(op);
+    auto interfaceOp = cast<PartitionableLoopsInterface>(*op);
+    auto partitionedLoops =
+        interfaceOp.getPartitionableLoops(kNumMaxParallelDims);
+    SmallVector<OpFoldResult> numThreads(numLoops, rewriter.getIndexAttr(0));
+    int64_t id = 0;
+    int64_t threadId = 0;
+    SmallVector<int64_t> blockTileSize = getTileSizes(op, 0);
+    SmallVector<int64_t> idDims;
+    for (unsigned loop : llvm::reverse(partitionedLoops)) {
+      if (loop >= blockTileSize.size()) continue;
+      int64_t num = elementPerWorkgroup[id++];
+      if (num > 1) {
+        numThreads[loop] = rewriter.getIndexAttr(num);
+        idDims.push_back(threadId++);
+      }
+    }
+    std::reverse(idDims.begin(), idDims.end());
+    for (int64_t i = threadId; i < 3; i++) idDims.push_back(i);
+
+    auto tilingResult =
+        linalg::tileToForeachThreadOp(rewriter, tilingOp, numThreads, idDims);
```

Comment:
Thanks for writing in a composable way.

It would be good to think about how to turn lines 102 - 127 into a separable transform that takes an op and returns a list of sizes to tile with.

Reply:
not for this PR obviously

Reply:
makes sense, right now the part 102-127 relies on attributes, but eventually we should get to something we can upstream